File: python/src/runapi/volcengine_lip_sync/resources/lip_sync_video.py
"""Volcengine Lip Sync video resource."""

from __future__ import annotations

from typing import Any, Optional

from runapi.core import Resource, RequestOptions

from ..contract_gen import CONTRACT
from ..types import CompletedLipSyncVideoResponse, LipSyncVideoResponse


class LipSyncVideo(Resource):
    """Drive a source video's lip movement from an audio track."""

    ENDPOINT = "/api/v1/volcengine_lip_sync/lip_sync_video"

    RESPONSE_CLASS = LipSyncVideoResponse
    COMPLETED_RESPONSE_CLASS = CompletedLipSyncVideoResponse

    def run(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Create a task and poll until it completes."""
        task = self.create(options=options, **params)
        return self._poll_until_complete(lambda: self.get(task.id, options=options))

    def create(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Create a lip-sync video task and return immediately with an ``id``."""
        compacted = self._compact_params(params)
        self._validate_contract(CONTRACT["lip-sync-video"], compacted)
        return self._request("post", self.ENDPOINT, body=compacted, options=options)

    def get(self, id: str, options: Optional[RequestOptions] = None) -> Any:
        """Fetch the current status of a lip-sync video task."""
        return self._request("get", f"{self.ENDPOINT}/{id}", options=options)

File: ruby/lib/runapi/volcengine_lip_sync/resources/lip_sync_video.rb
# frozen_string_literal: true

module RunApi
  module VolcengineLipSync
    module Resources
      # Volcengine Lip Sync video resource.
      class LipSyncVideo
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/volcengine_lip_sync/lip_sync_video"
        RESPONSE_CLASS = Types::LipSyncVideoResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedLipSyncVideoResponse

        def initialize(http)
          @http = http
        end

        # Create a lip-sync video task and wait until complete.
        #
        # @param params [Hash] lip-sync parameters
        # @return [RunApi::VolcengineLipSync::Types::CompletedLipSyncVideoResponse] completed task with videos
        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        # Start a lip-sync video task.
        #
        # @param params [Hash] lip-sync parameters
        # @return [RunApi::VolcengineLipSync::Types::LipSyncVideoResponse] task creation result with id
        def create(options: nil, **params)
          params = compact_params(params)
          validate_contract!(CONTRACT["lip-sync-video"], params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        # Get lip-sync video task status by task ID.
        #
        # @param id [String] task ID
        # @return [RunApi::VolcengineLipSync::Types::LipSyncVideoResponse] current task status
        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end
      end
    end
  end
end
